gateway-sp-comms/src/single_sp/startup.rs

```diff
@@ -0,0 +1,228 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2022 Oxide Computer Company
+
+//! Helper functionality for async startup of a `SingleSp`.
+
+use super::Inner;
+use super::InnerCommand;
+use crate::error::StartupError;
+use crate::SwitchPortConfig;
+use futures::Future;
+use gateway_messages::SpPort;
+use once_cell::sync::OnceCell;
+use slog::info;
+use slog::Logger;
+use std::net::SocketAddrV6;
+use std::time::Duration;
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc;
+use tokio::sync::watch;
+
+// States we can be in to allow immediate creation of a `SingleSp` even though
+// it can't immediately communicate on the management network.
+#[derive(Debug)]
+enum StartupState {
+    // We're waiting for our configured network interface to exist.
+    WaitingForInterface(String),
+    // We're waiting for the result of attempting to bind to our listening
+    // interface.
+    WaitingToBind(SocketAddrV6),
+    // Startup is complete (although not necessarily successful!).
+    //
+    // TODO We currently treat failure to bind a socket as a terminal error,
+    // expecting it to be indicative of a configuration problem we have no
+    // recourse to solve. Is that correct, or do we need some way of recovering
```

Comment:
I think this is fine. @Nieuwejaar what do you think?

Reply:
I believe this is true, if `WaitingForInterface` is satisfied by the appearance of an `ipadm` interface.  If that condition only means that there is a `dladm` link, then this isn't necessarily correct as you could be in a window between the creation of the link and the interface.  I suspect the condition is referring to the `ipadm`- interface, so this should be fine.

Reply:
I believe it is. The mechanics here are that we wait polling `if_nametoindex()` from libc until the interface exists. I just checked on a helios system: an interface existing in `dladm` only is not sufficient for `if_nametoindex()` to succeed; it needs to exist in `ipadm`.